### ISSUE TITLE
Do more releases?

### ISSUE THREAD

**Issue by Jarred-Sumner** [ziglang/zig#14542]
This is a proposal suggesting to release new versions of Zig more frequently.

Most projects written in Zig target the `master` build of zig, but breaking changes happen often. 

Breaking changes make it harder for people to contribute to open-source projects written in Zig. Instead of working on the project, potential contributors get frustrated figuring out how to get the project to compile. 

The bigger issue is not that Zig has breaking changes. The problem is: it's hard to know what version of Zig is expected to be used and how to download & install the project's supported version[^0]

**Why don't more projects use versioned releases of Zig?**

Zig's rapid progress is not reflected in the 3-6 month release cycle. You guys fix tons of bugs between releases.

Version | Release cycle
|-----:|-----|
| 0.10.0 - 0.10.1 | 101 days | 
| 0.9.1 - 0.10.0 | 260 days |
| 0.9.0 - 0.9.1 | 105 days |

If Zig had more frequent releases, like every 2 weeks, it would be easy for projects using Zig to say

> **This project expects Zig v0.123.456**

Package managers like homebrew, AUR, and elsewhere would have the expected version of Zig readily installable.

An idea related to #14475: now that Zig has a package manager, one could imagine Zig itself could download/use the expected version of Zig in a subcommand within `zig build` if prebuilt binaries for the host platform/arch/os/libc are available. Zig could be it's own version manager.

[^0]: It's great that binary builds are available (though it's missing a browseable index) and version managers like [zigup](https://github.com/marler8997/zigup) are nice, a commit sha technically works, but in practice &mdash; people have a hard time with this. They want to `brew install zig`. They don't want to clone zig's repo, read the instructions on compiling for macOS with CMake, install the specific matching version of llvm, etc

**Comment by michal-z** [comment]
One of the solutions is to check compiler version in `build.zig` and print error message that the compiler is too old. This is what I'm doing in zig-gamedev and it works pretty well: https://github.com/michal-z/zig-gamedev/blob/main/build.zig#L251

**Comment by Vexu** [comment]
This has been rejected previously in https://github.com/ziglang/zig/issues/4504#issuecomment-588299769

**Comment by phase** [comment]
A semi-solution to this could be frequent tagging of _release candidates_ that the community expects will have severe breaking changes or regressions (pre-1.0.0). Tagging `0.11.0-rc1` could let people install a bleeding edge build (without building from source) while communicating it's not a stable release.

**Comment by leecannon** [comment]
@phase the problem with that is breaking changes are not expected from one release candidate to the next nor from release candidates to the final release
and `let people install a bleeding edge build (without building from source)` is already satisfied by the daily master binaries on the website

**Comment by andrewrk** [comment]
I don't have any plans to change the release cadence in the next 2 years.

There will be many, many breaking changes in between now and 1.0. This is why I suggest for large projects such as Bun to stick to releases rather than master branch. I understand that you wanted to gain the performance benefits of master branch code, but when you do this, you opt into API breakage, potential regressions in master branch that would have been resolved by the time a release is tagged, and you have to start caring about small differences in zig versions.

> one could imagine Zig itself could download/use the expected version of Zig in a subcommand within `zig build` if prebuilt binaries for the host platform/arch/os/libc are available. Zig could be it's own version manager.

This is an open proposal: #10204

**Comment by marler8997** [comment]
I relate to the issues Jarred mentions and I think others who work with enough zig projects will also.  These issues are currently being exacerbated with a recent influx of breaking changes. Even though these growing pains won't last forever, I believe there are long-term benefits to be had from investing in relieving these issues now.  It means more projects can move forward early on which helps guide Zig by providing more real world usage.

Having more releases has the downside of adding extra work, but, I think we could improve the status quo without adding extra work for Andrew and the maintainers of Zig. I've created this repo in the hopes that it could help the situation:

https://github.com/marler8997/zig-unofficial-releases

I've listed one version so far, but, the idea is to pick versions every so often to recommend projects use based on when breaking changes come in.  It can also be used to document these breaking changes for us "zig explorers" to reference.  I've made it a repo so we can leverage github's PR system to propose which versions to "unofficially release".
